Add metadata token support

thibmeu opened the pull request:
Implementation for Publicly Verifiable Tokens for:
* The PrivateToken HTTP Authentication Scheme Extensions Parameter - [draft 1](https://datatracker.ietf.org/doc/html/draft-wood-privacypass-auth-scheme-extensions-01)
* Public Metadata Issuance - [draft 3](https://datatracker.ietf.org/doc/html/draft-hendrickson-privacypass-public-metadata-03)

The PR focuses on Publicly verifiable tokens with metadata. Private verifiable tokens present a similar interface, and would overload this PR.
This change introduces:
* A new example to use Publicly verifiable tokens with metadata
* A new class `Origin` that is used to verify tokens
* Two classes `Extension` and `Extensions`. They can be serialized as part of a header, as a TokenRequest, and instanciated by any party where required
* Three new classes `IssuerWithMetadata`, `ClientWithMetadata`, `OriginWithMetadata` that expose similar methods as their non metadata counterpart, enforcing passing extensions
* Test vectors that are to be submitted to the draft

This change depends on cloudflare/blindrsa-ts#15 to be merged. In the meantime, I've packaged cloudflare/blindrsa-ts as tgz.

armfazh commented on examples/pub_verif.example.ts:
```diff
@@ -44,7 +44,7 @@ async function rsaVariant(mode: BlindRSAMode): Promise<void> {
     //     |<-- Request+Token ---+                   |           |
     //     |                     |                   |           |
-    const isValid = await /*Origin*/ verifyToken(mode, token, issuer.publicKey);
+    const isValid = await /*Origin*/ new Origin(mode).verify(token, issuer.publicKey);
```
Create Origin with this info:
` const originInfo = ['origin.example.com', 'origin2.example.com'];`

armfazh commented on examples/pub_verif.example.ts:
```diff
@@ -12,7 +12,7 @@ async function rsaVariant(mode: BlindRSAMode): Promise<void> {
 
     // [ Issuer ] creates a key pair.
-    const keys = await keyGen(mode, {
+    const keys = await Issuer.generateKey(mode, {
```
let's create at the `Setup` step all the actors: Issuer, Origin, Client

thibmeu replied:
updated. I've added a `setup` function

armfazh commented on examples/pub_verif_metadata.example.ts:
```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2023 Cloudflare, Inc.
```
```suggestion
// Copyright (c) 2024 Cloudflare, Inc.
```

armfazh commented on examples/pub_verif_metadata.example.ts:
```diff
@@ -0,0 +1,75 @@
+async function rsaVariant(mode: PartiallyBlindRSAMode): Promise<void> {
+    // Protocol Setup
+    //
+    // [ Everybody ] agree to use Public Verifiable Tokens.
```
```suggestion
    // [ Everybody ] agree to use Public Verifiable Tokens with Metadata.
```

armfazh commented on examples/pub_verif_metadata.example.ts:
```diff
@@ -0,0 +1,75 @@
+    //     |<----- Request ------+                   |           |
+    const redemptionContext = crypto.getRandomValues(new Uint8Array(32));
+    const originInfo = ['origin.example.com', 'origin2.example.com'];
+    const tokChl = new TokenChallenge(tokenType, issuer.name, redemptionContext, originInfo);
```
call here to `origin.createTokenChallenge()`

thibmeu replied:
updated

armfazh commented on src/auth_scheme/private_token.ts:
```diff
@@ -285,6 +285,97 @@ export class Token {
     }
 }
 
+export type ExtensionType = number;
+
```
```suggestion
// enum {
//     reserved(0),
//     (65535)
// } ExtensionType;
export type ExtensionType = number;

```

thibmeu replied:
updated

armfazh commented on src/auth_scheme/private_token.ts:
```diff
@@ -285,6 +285,97 @@ export class Token {
+
+    serialize(): Uint8Array {
+        const output = new Array<ArrayBuffer>();
+
```
here you need to serialize in a `uint16` the number of extension in the array

thibmeu replied:
updated.

armfazh commented on src/auth_scheme/private_token.ts:
```diff
@@ -285,6 +285,97 @@ export class Token {
+    //     Extension extensions<0..2^16-1>;
+    // } Extensions;
+
+    constructor(public extensions: Extension[]) {}
```
we should enforce this 
`The list MUST be ordered by ExtensionType value, from 1 to 65535. `

armfazh commented on src/auth_scheme/private_token.ts:
```diff
@@ -388,17 +479,21 @@ export class WWWAuthenticateHeader {
-    constructor(public token: Token) {}
+    constructor(
+        public token: Token,
+        public extensions?: unknown,
```
```suggestion
        public extensions?: Extensions[],
```

armfazh commented on src/auth_scheme/private_token.ts:
```diff
@@ -388,17 +479,21 @@ export class WWWAuthenticateHeader {
         const attributes = data.split(',');
         let ppToken: Token | undefined = undefined;
-
+        let extentions: Extensions | undefined = undefined;
```
```suggestion
        let extentions: Array<Extensions>;
```

thibmeu replied:
the idea of using `Extensions|undefined` is to distinguish between token without extensions and token with extensions but with an empty array.

armfazh commented on src/pub_verif_token.ts:
```diff
@@ -1,11 +1,18 @@
 // Licensed under the Apache-2.0 license found in the LICENSE file or at https://opensource.org/licenses/Apache-2.0
 
-import { type BlindRSA, type BlindRSAPlatformParams, RSABSSA } from '@cloudflare/blindrsa-ts';
+import {
```
now looking at the diff of changes, I 'm favor of having a separate file for the code of  pub_verif_metatada.

thibmeu replied:
still with the abstract class, given the code is similar?

armfazh replied:
one file is ok

armfazh commented on src/pub_verif_token.ts:
```diff
@@ -83,33 +115,37 @@ export class TokenRequest {
-    static deserialize(bytes: Uint8Array): TokenRequest {
+    static deserialize(
+        bytes: Uint8Array,
+        tokenType: BlindRSAType | PartiallyBlindRSAType = BLIND_RSA,
```
let's not use default values in this function

thibmeu replied:
updated the code and its test.

armfazh commented on src/pub_verif_token.ts:
```diff
@@ -150,57 +216,132 @@ export class TokenResponse {
         public readonly params?: BlindRSAPlatformParams,
-    ) {}
+    ) {
+        this.suite = (extensions?: Extensions) => {
+            if (extensions === undefined) {
+                return BLIND_RSA.suite[this.mode]();
```
we can avoid this function just by returning the right suite.

```ts
if (!extensions){
  this.suite = BLIND_RSA.suite[this.mode]();
}else{
  const suite = PARTIALLY_BLIND_RSA.suite[this.mode]();
  const serializedExtensions = extensions.serialize();
  this.suite = {
    blindSign: (privateKey: CryptoKey, blindMsg: Uint8Array) =>
                        suite.blindSign(privateKey, blindMsg, serializedExtensions),
    verify: (publicKey: CryptoKey, signature: Uint8Array, message: Uint8Array) =>
                        suite.verify(publicKey, signature, message, serializedExtensions),
  };
}
```

thibmeu replied:
you don't know the extensions used for issuance at this point. the goal of this function is to provide the necessary suite function (`blindSign`, `verify`) so they don't need extension when called.

moreover, the issuer receiving the extensions as part of `ExtendedTokenRequest`, I've decided that the extensions would only be known at issuance time, and not upon creating the issuer. This is somehow closer to the spec. If extensions where to be passed at the constructor level, I'd argue it should be `supportedExtensions`, that would not have the value of the extension in them.

armfazh commented on src/pub_verif_token.ts:
```diff
@@ -150,57 +216,132 @@ export class TokenResponse {
+}
+
+export class Issuer extends PubliclyVerifiableIssuer {
+    constructor(
+        mode: BlindRSAMode | PartiallyBlindRSAMode,
+        name: string,
+        privateKey: CryptoKey,
+        publicKey: CryptoKey,
+        params?: BlindRSAPlatformParams,
+    ) {
+        super(mode, name, privateKey, publicKey, params);
+    }
```
if I'm not wrong, you can remove this constructor

thibmeu replied:
removed.

armfazh commented on src/pub_verif_token.ts:
```diff
@@ -150,57 +216,132 @@ export class TokenResponse {
+        super(mode, name, privateKey, publicKey, params);
+    }
+
+    async issue(tokReq: TokenRequest): Promise<TokenResponse> {
+        return super._issue(tokReq);
+    }
```
also you can make this method public in the abstract class, alternatively you can store the extension in the instance of the class

thibmeu replied:
The method has a different signature (due to types) for Issuer with metadata and issuer without.
Therefore the `_issue` method is private, and re-exported as `issue(TokenRequest|ExtendedTokenRequest)` where appropriate.

armfazh commented on src/pub_verif_token.ts:
```diff
@@ -247,10 +388,73 @@ export class Client {
+    private suite: Pick<BlindRSA, 'verify'>;
+
+    constructor(
+        public readonly mode: BlindRSAMode | PartiallyBlindRSAMode,
```
there is some `originInfo` defined in the examples than need to be set here.

thibmeu replied:
updated

armfazh reviewed (changes requested):
looks good in general, a few changes only

armfazh commented on examples/pub_verif.example.ts:
```diff
@@ -44,7 +50,7 @@ async function rsaVariant(mode: BlindRSAMode): Promise<void> {
     //     |<-- Request+Token ---+                   |           |
     //     |                     |                   |           |
-    const isValid = await /*Origin*/ verifyToken(mode, token, issuer.publicKey);
+    const isValid = await /*Origin*/ origin.verify(token, issuer.publicKey);
```
```suggestion
    const isValid = await origin.verify(token, issuer.publicKey);
```

armfazh commented on examples/pub_verif_metadata.example.ts:
```diff
@@ -0,0 +1,81 @@
+    // Protocol Setup
+    //
+    // [ Everybody ] agree to use Public Verifiable Tokens with Metadata.
+    const extensions = new Extensions([new Extension(0x0000, new Uint8Array([1, 2, 3]))]);
```
this is just to emphasize what metadata is and its packaging as a extension.
Also, the extension type cannot be 0.

```suggestion
    const metadata = new Uint8Array([1, 2, 3]);
    const extensions = new Extensions([new Extension(0xACDC, metadata)]);
```

thibmeu replied:
metadata is both the number and its content I'd say. I renamed the variable `info` instead
It might be worth to add grease reserved numbers to the draft, similar to what the main [Privacy Pass draft provides](https://www.ietf.org/archive/id/draft-ietf-privacypass-auth-scheme-15.html#section-6.2.1).
For now, I'll move `0xACDC` to a dedicated `TEST_EXTENSION_TYPE` to make this clear.

armfazh commented on src/auth_scheme/private_token.ts:
```diff
@@ -285,6 +286,123 @@ export class Token {
+        public extensionType: ExtensionType,
+        public extensionData: Uint8Array,
+    ) {
+        if (extensionType < 0 || extensionType >= MAX_UINT16) {
```
maybe we should also check that it is an integer number
```suggestion
        if (extensionType <= 0 || extensionType >= MAX_UINT16) {
```

thibmeu replied:
0 is allowed, even though it's reserved.
I've added `extensionType % 1 !== 0` though, to check it's an integer.

armfazh commented on examples/pub_verif_metadata.example.ts:
```diff
@@ -0,0 +1,81 @@
+    const token = await client.finalize(tokRes);
+    //     |<-- Request+Token ---+                   |           |
+    //     |                     |                   |           |
+    const isValid = await /*Origin*/ origin.verify(token, issuer.publicKey);
```
```suggestion
    const isValid = await origin.verify(token, issuer.publicKey);
```

armfazh commented on src/auth_scheme/private_token.ts:
```diff
@@ -285,6 +286,123 @@ export class Token {
+        public extensionData: Uint8Array,
+    ) {
+        if (extensionType < 0 || extensionType >= MAX_UINT16) {
+            throw new Error('invalid extension type. Must be between 0 and 2^16-1.');
```
```suggestion
            throw new Error('invalid value for extension type, MUST be from 1 to 2^16-1.');
```

thibmeu replied:
0x0000 is reserved. This does not mean it cannot be used

I've updated the message, even though I'm keeping `0` as the lower bound.

armfazh commented on src/auth_scheme/private_token.ts:
```diff
@@ -285,6 +286,123 @@ export class Token {
+    // } Extensions;
+
+    constructor(public extensions: Extension[]) {
+        let lastExtensionType = -1;
```
must check the number of extensions to be below MAXUINT16

thibmeu replied:
you're right. I've added a check

armfazh commented on src/auth_scheme/private_token.ts:
```diff
@@ -285,6 +286,123 @@ export class Token {
+    // } Extensions;
+
+    constructor(public extensions: Extension[]) {
+        let lastExtensionType = -1;
```
```suggestion
        let lastExtensionType = 0;
```

armfazh replied:
also not sure whether it is allowed to have more than one extension with duplicate types.

thibmeu replied:
0 is allowed. I could imagine extension of the same type.
I've opened https://github.com/chris-wood/draft-wood-privacypass-auth-scheme-extensions/pull/8

armfazh commented on src/auth_scheme/private_token.ts:
```diff
@@ -285,6 +286,123 @@ export class Token {
+        output.push(length);
+
+        for (const extension of this.extensions) {
+            const extensionBytes = extension.serialize();
+            let b = new ArrayBuffer(2);
+            new DataView(b).setUint16(0, extensionBytes.length);
+            output.push(b);
+            b = extensionBytes.buffer;
+            output.push(b);
```
```suggestion
            output.push(extension.serialize());
```

armfazh replied:
this is because the `Extension` type itself is already length-prefixed.
The prefix that was missing in the previous review was for `Extensions`.

thibmeu replied:
good catch. I've updated deserialization accordingly.

armfazh commented on src/auth_scheme/private_token.ts:
```diff
@@ -407,9 +529,18 @@ export class AuthorizationHeader {
+                case 'extensions': {
+                    if (tokenTypeEntry.publicMetadata) {
+                        const extEnc = base64url.parse(attrValue);
+                        extentions = Extensions.deserialize(extEnc);
```
does JS need break here?

```suggestion
                        extentions = Extensions.deserialize(extEnc);
                        break;
```

thibmeu replied:
it does

armfazh commented on test/pub_verif_token.test.ts:
```diff
@@ -12,12 +12,13 @@ import {
 
 // https://datatracker.ietf.org/doc/html/draft-ietf-privacypass-protocol-16#name-test-vectors
 import vectors from './test_data/pub_verif_v16.json';
+import { BLIND_RSA } from '../src/pub_verif_token.js';
```
```suggestion
```

thibmeu replied:
moved the import of `BLIND_RSA` to `const { BLIND_RSA, ... } = publicVerif`

thibmeu replied:
reading your comment after, I used `TOKEN_TYPES.BLIND_RSA` instead

armfazh commented on test/pub_verif_token.test.ts:
```diff
@@ -51,6 +52,18 @@ async function keysFromVector(v: Vectors): Promise<[CryptoKeyPair, Uint8Array]>
     return [{ privateKey, publicKey }, publicKeyEnc];
 }
 
+class TestSerializeTokenRequest {
```
no need to use this class if you use this call, 
```
testSerializeType(TOKEN_TYPES.BLIND_RSA, TokenRequest, tokReq);
```
and change in ./src/pub_verif_token.ts
```diff
-        tokenType: BlindRSAType | PartiallyBlindRSAType,
+        tokenType: TokenTypeEntry,
```
and validate the type in the constructor.

thibmeu replied:
we need a new class for typing reason. `deserialize` of TokenRequest requires a `TokenTypeEntry` parameter, and therefore the signature does not match the expected `CanDeserialize`
see the typescript error
```
Type '(bytes: Uint8Array, tokenType: TokenTypeEntry) => TokenRequest' is not assignable to type '(type: TokenTypeEntry, _b: Uint8Array) => CanSerialize'.
```

thibmeu replied:
EDIT: I've changed the function signature to be `(TokenTypeEntry, Uint8Array) => TokenRequest`, and this matches `testSerializeType` definition. Ordering needed to change.

armfazh commented:
In the latest doc, it seems it allows to pass a config file
`-c .gitguardian.yml`
so we can exclude `'test/test_data/*` from scanning,

https://github.com/GitGuardian/ggshield-action?tab=readme-ov-file#adding-extra-options-to-the-action

thibmeu commented:
> In the latest doc, it seems it allows to pass a config file `-c .gitguardian.yml` so we can exclude `'test/test_data/*` from scanning,
> 
> https://github.com/GitGuardian/ggshield-action?tab=readme-ov-file#adding-extra-options-to-the-action

I'm fine merging with a failing `GitGuardian`, and I'd prefer doing this (given it's a one off basis) rather than adding one more file at the root.

armfazh commented on src/auth_scheme/private_token.ts:
```diff
@@ -305,8 +305,10 @@ export class Extension {
     ) {
-        if (extensionType < 0 || extensionType >= MAX_UINT16) {
-            throw new Error('invalid extension type. Must be between 0 and 2^16-1.');
+        if (extensionType < 0 || extensionType >= MAX_UINT16 || extensionType % 1 !== 0) {
```
how about using https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger

extension type can be equal to MAX_UINT16?

thibmeu replied:
good catch about MAX_UINT.
using `Number.isInteger` should work as well. I somehow had some edge cases in mind, but looks like it's safe

armfazh commented on src/auth_scheme/private_token.ts:
```diff
@@ -285,6 +286,125 @@ export class Token {
+        offset += 2;
+
+        const extensions = new Array<Extension>();
+        for (let i = 0; i < extensionsLength && offset < bytes.length; i++) {
+            // retrieve the length of the extension's data
+            const len_data = input.getUint16(offset + 2);
+            // { type: uint16, len: uint16, data: opaque }
+            const len = 2 + 2 + len_data;
+            const extensionBytes = new Uint8Array(input.buffer.slice(offset, offset + len));
+            offset += len;
+            extensions.push(Extension.deserialize(extensionBytes));
```
you can avoid reading twice each extension length by returning the offset of the last deserialization.

To get the offset I passed an `ops` object, see this code as an example
https://github.com/cloudflare/privacypass-ts/blob/main/src/auth_scheme/private_token.ts#L196-L222

thibmeu replied:
done.
I'm not sure this is more efficient though, given `deserialize` takes a `Uint8Array`, and it therefore requires to pass `new Uint8Array(input.buffer.slice(offset))`, without slicing the end of the array.

armfazh commented on test/auth_scheme_with_extensions.test.ts:
```diff
@@ -0,0 +1,82 @@
+        expect(invalidExtensionsConstructor).toThrow();
+    });
+
+    test('Too many extensions', () => {
```
add a test case for duplicated extensions and another one for sorted extensions

armfazh commented on test/pub_verif_token.test.ts:
```diff
@@ -74,7 +74,8 @@ describe.each(vectors)('PublicVerifiable-Vector-%#', (v: Vectors) => {
         const client = new Client(mode);
         const tokReq = await client.createTokenRequest(tokChl, publicKeyEnc);
-        testSerialize(TokenRequest, tokReq);
+        // use a custom class to have only beign one parameter on the `deserialize` method
```
```suggestion
```

armfazh commented on src/pub_verif_token.ts:
```diff
@@ -247,10 +375,90 @@ export class Client {
+            this.suite = BLIND_RSA.suite[this.mode]();
+            this.tokenType = BLIND_RSA;
+        } else {
+            const suite = PARTIALLY_BLIND_RSA.suite[this.mode]();
+            const extensions = this.extensions.serialize();
+            this.suite = {
+                verify: (publicKey: CryptoKey, signature: Uint8Array, message: Uint8Array) =>
+                    suite.verify(publicKey, signature, message, extensions),
+            };
+            this.tokenType = PARTIALLY_BLIND_RSA;
```
Current code can be simplified. What if we store both the suite and the extensions, then just call `verify` with the right parameters.

thibmeu replied:
the extension is stored. However, at the verification step, the type of the suite is not known. We therefore cannot easily pass `extensions` parameter at that time.
Having this function wrapping in the constructor, which stores the extension and passes them when necesarry allows to provide consistent typing.

armfazh commented on src/pub_verif_token.ts:
```diff
@@ -247,10 +375,90 @@ export class Client {
+
+export class OriginWithMetadata extends PubliclyVerifiableOrigin {
+    constructor(
+        mode: BlindRSAMode | PartiallyBlindRSAMode,
```
```suggestion
        mode: PartiallyBlindRSAMode,
```

armfazh commented on src/pub_verif_token.ts:
```diff
@@ -247,10 +375,90 @@ export class Client {
+}
+
+export class Origin extends PubliclyVerifiableOrigin {
+    constructor(mode: BlindRSAMode | PartiallyBlindRSAMode, originInfo?: string[]) {
```
```suggestion
    constructor(mode: BlindRSAMode, originInfo?: string[]) {
```

armfazh commented on src/pub_verif_token.ts:
```diff
@@ -150,57 +213,122 @@ export class TokenResponse {
-export class Client {
+export class IssuerWithMetadata extends PubliclyVerifiableIssuer {
+    constructor(
+        mode: BlindRSAMode | PartiallyBlindRSAMode,
```
```suggestion
        mode: PartiallyBlindRSAMode,
```

armfazh replied:
For ` class Issuer extends PubliclyVerifiableIssuer`, one can add a constructor accepting `mode:BlindRSAMode` only.

thibmeu replied:
agree

thibmeu commented:
commits have been squashed and branch rebased on `main`

thibmeu commented:
squashed all commits ahead of merge

armfazh commented on src/auth_scheme/private_token.ts:
```diff
@@ -285,6 +286,143 @@ export class Token {
+            //
+            // struct {
+            //     Extension extensions<0..2^32-1>;
+            // } Extensions;
```
you can remove my comment, not required anymore

armfazh commented on test/auth_scheme_with_extensions.test.ts:
```diff
@@ -0,0 +1,162 @@
+        testSerialize(Extensions, arrayExt);
+
+        const arrayExtBytes = arrayExt.serialize();
+        expect(arrayExtBytes.length).toBe(
+            2 /* <-- This prefix should be at least 4 bytes. */ +
+                (2 + 2 + Extension.MAX_EXTENSION_DATA_LENGTH),
+        );
```
```suggestion
        expect(arrayExtBytes.length).toBe(
            2 /* Length-prefix for array of extensions. */ +
                (2 + 2 + Extension.MAX_EXTENSION_DATA_LENGTH),
        );
```

armfazh commented on test/auth_scheme_with_extensions.test.ts:
```diff
@@ -0,0 +1,162 @@
+
+    test('maxSizeArrayExtension', () => {
+        const ext = new Extension(0xaa, new Uint8Array(Extension.MAX_EXTENSION_DATA_LENGTH));
+        const arrayExt = new Extensions([ext]);
```
with the current changes, an array can only have one maximum-length extension, but not two or more, for example, this throws an error

```
const ext1 = new Extension(0xaa, new Uint8Array(Extension.MAX_EXTENSION_DATA_LENGTH));
const ext2 = new Extension(0xbb new Uint8Array(Extension.MAX_EXTENSION_DATA_LENGTH));
const arrayExt = new Extensions([ext1, ext2]);
arrayExt.serialize();
```

If this is the intended behaviour, we should warn in comments.

thibmeu replied:
it is the expected behaviour yes.
I've added a comment.

thibmeu replied:
it is the expected behaviour yes.
I've added a comment.

thibmeu committed:
Add Publicly Verifiable Token with Metadata

Draft 2 https://datatracker.ietf.org/doc/html/draft-hendrickson-privacypass-public-metadata-03#name-client-to-issuer-request-2

Using Partially Blind RSA implementation.
Given the protocol is close to the normal Public issuance, the class is
abstracted, and two children of this abstract class are exported: one
with Metadata and the other without.

There are no test vectors yet for the protocol, which is to be added in
a following commit.

Add Origin class for publicly verifiable token

Add example for metadata

Add test vectors for Privacy Pass with metadata

Not yet part of the draft, but allow to confirm issues with the
implementation.

Add Partially Blind RSA as supported token in README

thibmeu commented:
last commit to update @cloudflare/blindrsa-ts to 0.4.2
